src/main/java/com/google/devtools/build/lib/rules/repository/RepoRecordedInput.java

```diff
@@ -106,121 +137,192 @@ public int compareTo(RepoRecordedInput o) {
-                    Root.fromPath(path.getParentDirectory()),
-                    PathFragment.create(path.getBaseName())));
+            return new FileOutsideWorkspace(PathFragment.create(s));
           }
         };
 
     @Override
     public Parser getParser() {
       return PARSER;
     }
+
+    /**
+     * Convert to a {@link com.google.devtools.build.lib.actions.FileValue} to a String appropriate
+     * for placing in a repository marker file.
+     *
+     * @param fileValue The value to convert. It must correspond to a regular file.
+     */
+    public static String fileValueToMarkerValue(FileValue fileValue) throws IOException {
+      Preconditions.checkArgument(fileValue.isFile() && !fileValue.isSpecialFile());
+      // Return the file content digest in hex. fileValue may or may not have the digest available.
+      byte[] digest = fileValue.realFileStateValue().getDigest();
+      if (digest == null) {
+        // Fast digest not available, or it would have been in the FileValue.
+        digest = fileValue.realRootedPath().asPath().getDigest();
+      }
+      return BaseEncoding.base16().lowerCase().encode(digest);
+    }
   }
 
-  /** Represents a file input accessed during the repo fetch that is addressable by a label. */
-  public static final class LabelFile extends File {
-    final Label label;
+  /**
+   * Represents a file input accessed during the repo fetch that is within the current Bazel
+   * workspace.
+   *
+   * <p>This is <em>almost</em> like being addressable by a label, but includes the extra corner
+   * case of files inside a repo but not within any package due to missing BUILD files. For example,
+   * the file {@code @@foo//:abc.bzl} is addressable by a label if the file {@code @@foo//:BUILD}
+   * exists. But if the BUILD file doesn't exist, the {@code abc.bzl} file should still be
+   * "watchable"; it's just that {@code @@foo//:abc.bzl} is technically not a valid label.
+   */
+  public static final class FileInsideWorkspace extends File {
+    private final RepositoryName repoName;
+    private final PathFragment pathFragment;
+
+    public FileInsideWorkspace(RepositoryName repoName, PathFragment pathFragment) {
+      this.repoName = repoName;
+      this.pathFragment = pathFragment;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (!(o instanceof FileInsideWorkspace)) return false;
+      FileInsideWorkspace that = (FileInsideWorkspace) o;
+      return Objects.equals(repoName, that.repoName)
+          && Objects.equals(pathFragment, that.pathFragment);
+    }
 
-    public LabelFile(Label label) {
-      this.label = label;
+    @Override
+    public int hashCode() {
+      return Objects.hash(repoName, pathFragment);
     }
 
     @Override
-    String toStringInternal() {
-      return label.toString();
+    public String toStringInternal() {
+      // We store `@@foo//abc/def:ghi.bzl` as just `@@foo//abc/def/ghi.bzl`.
```

Comment:
nit: also tell why. It's easy to tell from the code what's happening, but it's impossible to tell why you decided that to be the case (or just say "see class javadoc")

Reply:
done.